### ISSUE TITLE
How can I "regenerate the last n tokens"

### ISSUE THREAD

**Issue by l3utterfly** [ggerganov/llama.cpp#5192]
I want to be able to regenerate the last N output tokens. Using `main` as an example, is there an easy way to do this without restarting from the beginning (loading from cached session tokens, etc.)?

My intuition says it should be something simple as adjusting the `n_past` and striking some values from the `kv_cache`. Can someone help point me in the right direction to implement this please?

**Comment by ggerganov** [comment]
```c
const int n_regen = 10; // change to fit your need

// clear the last n_regen tokens from the KV cache and update n_past
llama_kv_cache_seq_rm(ctx, 0, n_past - n_regen, -1);
n_past -= n_regen;
```

**Comment by ggerganov** [comment]
Yes, we might have to extend the sampling API to do such things.

There is a helper function for making copies of the `llama_sampling_context` which can be useful:

https://github.com/ggerganov/llama.cpp/blob/d0e10bf1b2617712dfb3a89b7ae264f77dde9a3d/common/sampling.h#L77-L79

This can resolve some use cases - for example, it is used in the `speculative` example to manage the draft sampling contexts. But it obviously won't work in all situations, especially in cases where you have to regenerate a dynamic number of tokens.

**Comment by l3utterfly** [comment]
I managed to implement the regenerate feature. The PR above was the APIs I needed for it.

I noticed even after rolling back the `prev` tokens in the sampler and decreasing `n_past`, I needed to run `llama_decode` again on the last token to update the logits.

Do you think we need to create an overarching API that handles both rolling back the sampler and calling decode on the last token? or should it be left as an exercise to the end-user.

**Comment by MaggotHATE** [comment]
> ```c
> llama_kv_cache_seq_rm(ctx, 0, n_past - n_regen, -1);
> ```

This may break on large output if group-attention is used, since n_past is resized when reaching grp_attn_w. Any message larger than grp_attn_w would "reset" n_past and break regenerating.

UPD: I'm not sure about [this part](https://github.com/ggerganov/llama.cpp/blob/44fbe34360dd760f9e68b4271f21533436397f84/examples/main/main.cpp#L563C17-L563C48) in group-attention
```C++
while (n_past >= ga_i + ga_w) {
```
Shouldn't it be `ga_i * ga_w` ? Or do I misunderstand how it works?

**Comment by ggerganov** [comment]
Yes, regeneration of more tokens than in the current "uncompressed" window should be asserted as invalid when using self-extend

**Comment by miaout17** [comment]
Is this functionality still working? I noticed that the PR was merged but some of the code disappeared later. 

I am trying the simplest thing to go back by one token by doing this in the loop in the main.cpp example. 
```
// My input is 16 tokens, so this hacky check ensures that only output generation
// is reverted.
if (embd.size() == 1) {
  llama_kv_cache_seq_rm(ctx, 0, n_past, -1);
  common_sampler_reset(smpl);
}
```

I'm expecting to seeing similar tokens in each step (it should stuck at generating the 1st token after input). However I'm surprised to find that it seems to remember the last few tokens. Am I doing anything wrong?

**Comment by miaout17** [comment]
Please disregard my last comment. I just realized it's simply this...

> I noticed even after rolling back the prev tokens in the sampler and decreasing n_past, I needed to run llama_decode again on the last token to update the logits.

Feeding the last token into the batch and calling `llama_decode` solved the problem.
